File: app/blueprints/community_map.py
from flask import Blueprint, render_template, request, session, jsonify
from app.models.user import User, db
from app.models.community_pin import CommunityPin

community_map_bp = Blueprint('community_map', __name__, template_folder='../templates')

@community_map_bp.route('/community-map', methods=['GET'])
def view_community_map():
    if 'user_id' not in session:
        return jsonify({"error": "Unauthorized"}), 401
    user = User.query.get(session['user_id'])
    return render_template('community_map.html', user=user)

@community_map_bp.route('/api/community/pins', methods=['GET'])
def get_global_pins():
    if 'user_id' not in session: return jsonify({"error": "Unauthorized"}), 401
    
    pins = CommunityPin.query.order_by(CommunityPin.created_at.desc()).all()
    
    pins_data = []
    for p in pins:
        pins_data.append({
            "id": p.id,
            "name": p.spot_name,
            "lat": p.latitude,
            "lng": p.longitude,
            "species": p.species or "N/A",
            "lure": p.lure_used or "N/A",
            "season": p.time_of_year or "Unknown",
            "creator_name": p.creator.display_name or p.creator.username if p.creator else "Anonymous", # Pull username
            "conditions": {
                "pressure": p.pressure or "N/A",
                "cloud": p.cloud_cover or "N/A",
                "rain": p.rain or "N/A",
                "temp": p.temp or "N/A",
                "wind": p.wind or "N/A"
            }
        })
    return jsonify({"pins": pins_data})

@community_map_bp.route('/api/community/pins/create', methods=['POST'])
def save_community_pin():
    if 'user_id' not in session: return jsonify({"error": "Unauthorized"}), 401
    uid = session['user_id']
    data = request.get_json() or {}
    
    name = data.get('spot_name', '').strip()
    lat = data.get('latitude')
    lng = data.get('longitude')
    
    if not name or lat is None or lng is None:
        return jsonify({"error": "Missing pin coordinates or location name parameters."}), 400
        
    p = CommunityPin(
        user_id=uid, 
        spot_name=name,
        latitude=float(lat),
        longitude=float(lng),
        species=data.get('species'),
        lure_used=data.get('lure_used'),
        time_of_year=data.get('time_of_year'),
        pressure=data.get('pressure'),
        cloud_cover=data.get('cloud_cover'),
        rain=data.get('rain'),
        temp=data.get('temp'),
        wind=data.get('wind')
    )
    db.session.add(p)
    db.session.commit()
    return jsonify({"success": True, "pin_id": p.id})

@community_map_bp.route('/api/community/pins/delete/<int:pin_id>', methods=['DELETE'])
def admin_delete_pin(pin_id):
    if 'user_id' not in session: return jsonify({"error": "Unauthorized"}), 401
    
    if int(session['user_id']) != 1:
        return jsonify({"error": "Forbidden: Only administrative master profiles can clear community map pins."}), 403
        
    p = CommunityPin.query.get(pin_id)
    if not p:
        return jsonify({"error": "Target pin location does not exist inside global logs."}), 404
        
    db.session.delete(p)
    db.session.commit()
    return jsonify({"success": True})

File: static/js/community_map.js
let map = null;
let activePinsCache = [];
let leafletMarkersGroup = null;
let currentExpandedPinId = null;

const currentClientUserId = parseInt("{{ user.id }}", 10);

function initLeafletCanvas() {
    map = L.map('mapViewport', { attributionControl: false }).setView([42.08, -87.73], 10);
    
    L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
        maxZoom: 19
    }).addTo(map);

    leafletMarkersGroup = L.layerGroup().addTo(map);

    map.on('click', function(e) {
        openLogModal(e.latlng.lat, e.latlng.lng);
    });
}

async function pullMapPinsPayload() {
    const res = await fetch('/api/community/pins');
    const data = await res.json();
    activePinsCache = data.pins || [];
    
    syncPinsToCanvas();
    renderSidebarPinsList();
}

function syncPinsToCanvas() {
    leafletMarkersGroup.clearLayers();
    
    activePinsCache.forEach(p => {
        const marker = L.circleMarker([p.lat, p.lng], {
            radius: 8,
            fillColor: '#E5509B',
            color: '#ffffff',
            weight: 2,
            opacity: 1,
            fillOpacity: 0.9
        });

        marker.on('click', (e) => {
            L.DomEvent.stopPropagation(e);
            focusPinNode(p.id, true);
        });

        marker.addTo(leafletMarkersGroup);
        p._markerRef = marker;
    });
}

function renderSidebarPinsList() {
    const q = document.getElementById('pinSearchField').value.toLowerCase();
    const deck = document.getElementById('pinsListDeck');
    deck.innerHTML = '';

    const filtered = activePinsCache.filter(p => 
        p.name.toLowerCase().includes(q) || 
        p.species.toLowerCase().includes(q) || 
        p.creator_name.toLowerCase().includes(q) ||
        p.lure.toLowerCase().includes(q)
    );

    if(filtered.length === 0) {
        deck.innerHTML = '<div style="color:#555; padding:20px; text-align:center; font-size:0.85rem;">No shared coordinates matched query criteria.</div>';
        return;
    }

    filtered.forEach(p => {
        const isExpanded = p.id === currentExpandedPinId;
        const card = document.createElement('div');
        card.id = `sidebar-card-${p.id}`;
        card.style = `padding:15px; border-bottom:1px solid #1e1e1e; cursor:pointer; background:${isExpanded ? '#1e1e1e' : 'transparent'}; transition: background 0.2s;`;
        card.onclick = () => focusPinNode(p.id, false);
        
        card.innerHTML = `
            <div style="display:flex; justify-content:space-between; align-items:center;">
                <strong style="color:${isExpanded ? 'var(--primary)' : '#fff'}; font-size:0.95rem;">📍 ${p.name}</strong>
                <span style="color:#666; font-size:0.7rem; font-family:monospace;">By: ${p.creator_name}</span>
            </div>
            
            ${isExpanded ? `
                <div style="margin-top:12px; display:flex; flex-direction:column; gap:6px; font-size:0.8rem; color:#ccc; border-top:1px solid #2d2d2d; padding-top:10px;">
                    <div>🐟 <span style="color:#888;">Target Species:</span> <strong>${p.species}</strong></div>
                    <div>🪱 <span style="color:#888;">Lure Profile:</span> <strong>${p.lure}</strong></div>
                    <div>📅 Seasonality: <strong>${p.season}</strong></div>
                    <div style="background:#121212; padding:8px; border-radius:6px; margin-top:4px; display:grid; grid-template-columns:1fr 1fr; gap:6px; font-size:0.75rem;">
                        <div>⏱️ Baro: <span style="color:#fff;">${p.conditions.pressure}</span></div>
                        <div>☁️ Cloud: <span style="color:#fff;">${p.conditions.cloud}</span></div>
                        <div>🌧️ Rain: <span style="color:#fff;">${p.conditions.rain}</span></div>
                        <div>🌡️ Temp: <span style="color:#fff;">${p.conditions.temp}</span></div>
                        <div style="grid-column: 1 / span 2;">💨 Wind: <span style="color:#fff;">${p.conditions.wind}</span></div>
                    </div>
                    
                    
                    ${currentClientUserId === 1 ? `
                        <button onclick="dispatchPinDeletion(${p.id}, event)" style="margin-top:10px; width:100%; border:1px solid #f43f5e; background:none; color:#f43f5e; padding:6px; border-radius:4px; font-size:0.75rem; font-weight:bold; cursor:pointer;">Remove Spot Coordinates (Admin Override)</button>
                    ` : ''}
                </div>
            ` : ''}
        `;
        deck.appendChild(card);
    });
}

function focusPinNode(id, scrollSidebar) {
    currentExpandedPinId = id;
    const pin = activePinsCache.find(p => p.id === id);
    if(!pin) return;

    map.panTo([pin.lat, pin.lng]);
    renderSidebarPinsList();

    if(scrollSidebar) {
        setTimeout(() => {
            const el = document.getElementById(`sidebar-card-${id}`);
            if(el) el.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
        }, 500);
    }
}

function openLogModal(lat, lng) {
    document.getElementById('modalLatField').value = lat;
    document.getElementById('modalLngField').value = lng;
    document.getElementById('modalCoordPreview').value = `${lat.toFixed(4)}, ${lng.toFixed(4)}`;
    
    document.getElementById('modalNameField').value = '';
    document.getElementById('modalSpeciesField').value = '';
    document.getElementById('modalLureField').value = '';
    document.getElementById('modalSeasonField').value = '';
    document.getElementById('modalBaroField').value = '';
    document.getElementById('modalCloudField').value = '';
    document.getElementById('modalRainField').value = '';
    document.getElementById('modalTempField').value = '';
    document.getElementById('modalWindField').value = '';

    document.getElementById('pinLogModal').style.display = 'flex';
}

function closeModal() {
    document.getElementById('pinLogModal').style.display = 'none';
}

async function dispatchPinCreation(e) {
    e.preventDefault();
    const payload = {
        latitude: document.getElementById('modalLatField').value,
        longitude: document.getElementById('modalLngField').value,
        spot_name: document.getElementById('modalNameField').value.trim(),
        species: document.getElementById('modalSpeciesField').value.trim(),
        lure_used: document.getElementById('modalLureField').value.trim(),
        time_of_year: document.getElementById('modalSeasonField').value.trim(),
        pressure: document.getElementById('modalBaroField').value.trim(),
        cloud_cover: document.getElementById('modalCloudField').value.trim(),
        rain: document.getElementById('modalRainField').value.trim(),
        temp: document.getElementById('modalTempField').value.trim(),
        wind: document.getElementById('modalWindField').value.trim()
    };

    const res = await fetch('/api/community/pins/create', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
    });

    if(res.ok) {
        closeModal();
        await pullMapPinsPayload();
        const freshData = await res.json();
        focusPinNode(freshData.pin_id, true);
    } else {
        alert("Failed to publish coordinate parameters onto the global map canvas.");
    }
}

async function dispatchPinDeletion(id, event) {
    event.stopPropagation();
    if(!confirm("ADMIN SECURITY OVERRIDE: Remove this public spot permanently across all community view boards?")) return;

    const res = await fetch(`/api/community/pins/delete/${id}`, { method: 'DELETE' });
    if(res.ok) {
        currentExpandedPinId = null;
        await pullMapPinsPayload();
    } else {
        const errorMsg = await res.json();
        alert(errorMsg.error || "Administrative rejection encountered.");
    }
}

document.addEventListener("DOMContentLoaded", () => {
    initLeafletCanvas();
    pullMapPinsPayload();
});
